### PR TITLE
Improve previous line count - fixes #12

### LINKED ISSUE #12: improve previous line count
The previous line count is being tracked by `log-update` but it does not take the terminal size into account.

For instance, if you log a line of 40 characters while the terminal is only 30 characters, it will be printed on 2 lines. After calling `log-update` again, it will only erase the 2nd line because the previous line count was only 1 and not 2.

So we have to iterate over all the lines [here](https://github.com/sindresorhus/log-update/blob/master/index.js#L12) and calculate the number of printed lines instead.

### PR DESCRIPTION
**Pull request by SamVerschueren**
It seems to be as simple as this. Tested it a couple of times, also while resizing the window and it worked perfectly.

Some things I might have missed?

### CODE REVIEW

**Comment by kevva** [comment]
lgtm :+1:

**Comment by SamVerschueren** [comment]
Friendly ping :). Just check out when you find the time.

**Comment by sindresorhus** [comment]
Before this PR:

before

After this PR:

after

**Comment by SamVerschueren** [comment]
Odd, will look into it.

**Comment by SamVerschueren** [comment]
Alright, I found out why this fails. Odd that it worked in HyperTerm though.

I use `wrap-ansi` here, but the thing is that it doesn't wrap like a terminal wraps a string. The terminal doesn't split on a space, it just breaks the word in 2.

`wrap-ansi` on the other hand splits words differently. This is what it looks if I use `35` columns.

> [1]  0.9565833098848577 ~
> 0.1412900921566833
> [2]  0.8579648462764315

While it should've been

> [1]  0.9565833098848577 ~ 0.1412900
> 921566833
> [2]  0.8579648462764315

**Comment by sindresorhus** [comment]
@SamVerschueren Ah, yeah. Try setting the `hard` option in `wrap-ansi`.

**Comment by SamVerschueren** [comment]
Not working either because the two words still fit on two lines (one word per line). This only works if one word exceeds the length of the entire line.

**Comment by sindresorhus** [comment]
@SamVerschueren Ok, ideally that module should support an option for wrapping on just plain text without caring about words (wanna open an issue there about that?), but for now, I guess we can use https://github.com/chalk/slice-ansi instead.

**Comment by SamVerschueren** [comment]
🎉  https://github.com/chalk/wrap-ansi/issues/12

Maybe I could look into adding the word-wrapping option to `wrap-ansi` instead of working around the problem temporarily with `slice-ansi`?

**Comment by sindresorhus** [comment]
Yes, that would definitely be the best solution, but might be more work, so I didn't want to ask you to do that.

**Comment by SamVerschueren** [comment]
Will look into it when I find some time. Then I can still decide what to do with it :).

**Comment by SamVerschueren** [comment]
I finally came into finalizing this.

Apparently Hyper and iTerm behave differently. When `iTerm` renders a long string on two lines, you'd only have to erase 1 line. In `Hyper` however, when you render a long string on two lines, you would have to erase 2 lines.

So before rendering the data to the screen, I use the updated `wrap-ansi` where I added the [`wordWrap`](https://github.com/chalk/wrap-ansi/pull/13) option. This way, we will always render 2 lines (iTerm and Hyper) because the line is split by ourselves.

**Comment by sindresorhus** [comment]
> Apparently Hyper and iTerm behave differently. When iTerm renders a long string on two lines, you'd only have to erase 1 line. In Hyper however, when you render a long string on two lines, you would have to erase 2 lines.

Can you check some other terminals? If it's only Hyper doing that, it's probably a bug.

**Comment by SamVerschueren** [comment]
The integrated terminal of VS Code behaves the same as Hyper. So does the command prompt and powershell CLI on Windows.

**Comment by sindresorhus** [comment]
> The integrated terminal of VS Code behaves the same as Hyper.

No surprise there. They use the same terminal rendering package.

File: index.js [hunk 1]
```diff
@@ -1,13 +1,16 @@
 'use strict';
 const ansiEscapes = require('ansi-escapes');
 const cliCursor = require('cli-cursor');
+const termSize = require('term-size');
```
Comment (sindresorhus) [inline index.js]: You don't really need `term-size` here as we're only doing this interactively, so we can just use `process.stdout.columns`.

**Comment by sindresorhus** [comment]
I still think it's a bug in those terminals, but gonna merge since it fixes many problems. When the terminal wraps a line it still count as one line and should only require going one line up.

@qix- Any ideas?

**Comment by SamVerschueren** [comment]
According to issue #12, this isn't working properly on Windows either. Haven't really find the time to test and debug it.

**Comment by astoilkov** [comment]
Have you looked into [draftlog](https://github.com/ivanseidel/node-draftlog)? I think they cache the non visible lines and when they come into view they update them.

**Comment by SamVerschueren** [comment]
@astoilkov I guess you wanted to answer #10 ?

**Comment by Qix-** [comment]
> Apparently Hyper and iTerm behave differently. When iTerm renders a long string on two lines, you'd only have to erase 1 line. In Hyper however, when you render a long string on two lines, you would have to erase 2 lines.

That actually sounds like iTerm is behaving incorrectly. Let me test.

**Comment by SamVerschueren** [comment]
I used this little script to test the behaviour.

```js
'use strict';
const ansiEscapes = require('ansi-escapes');

console.log('This is a piece of text that should be rendered on two lines. In iTerm, this text will be rendered only once. In Hyper however, you will see it 1,5 times.');

console.log('This is a piece of text that should be rendered on two lines. In iTerm, this text will be rendered only once. In Hyper however, you will see it 1,5 times.');

process.stdout.write(ansiEscapes.eraseLines(2));
```

Set the viewport to approximately 90 characters. The text is rendered twice and then the 2nd console.log is erased (1 empty line + the entire 1 line of text). This works in iTerm but not in Hyper because in Hyper renders one console.log as 2 separate lines. So to make it work in Hyper, you would have to use `ansiEscapes.eraseLines(3)`

**Comment by Qix-** [comment]
Thanks, that did it.

Seems like `ansi-escapes` is using hack-ish escapes anyway.

```console
$ node test-ansi.js | cat -v
This is a piece of text that should be rendered on two lines. In iTerm, this text will be rendered only once. In Hyper however, you will see it 1,5 times.
This is a piece of text that should be rendered on two lines. In iTerm, this text will be rendered only once. In Hyper however, you will see it 1,5 times.
^[[1000D^[[K^[[1A^[[1000D^[[K
```

- `^[[1000D` is a hack; `^[[G` should be used instead.
- Using `^[[G` instead of `^[[1000D` actually produced the 1.5 result in both terminals.
- Instead of any of the above sequences, using `^[[2A^[[G^[[J` (move 2 up, absolute left, erase to end of display) worked as expected

**Comment by sindresorhus** [comment]
> However, I'm pretty convinced it's due to that hack-ish 1000D escape.

Does it still not work with the change to `^[[G`?

---

I've read through this thread multiple times now, but it's not really clear what needs to change or if something needs to change at all?

**Comment by SamVerschueren** [comment]
It's just a Windows issue. Haven't looked into it though and it might be caused by `wrap-ansi`, see #19

**Pull request merged**